modules/openapi-generator/src/main/resources/scala-sttp/api.mustache

```diff
@@ -4,42 +4,35 @@ package {{package}}
 import {{import}}
 {{/imports}}
-import {{invokerPackage}}._
-import alias._
+import {{invokerPackage}}.JsonSupport._
 import sttp.client._
 import sttp.model.Method
 
 {{#operations}}
-object {{classname}} {
-
-  def apply(baseUrl: String = "{{{basePath}}}")(implicit serializer: SttpSerializer) = new {{classname}}(baseUrl)
-}
-
-class {{classname}}(baseUrl: String)(implicit serializer: SttpSerializer) {
-
-  import Helpers._
-  import serializer._
+class {{classname}}(baseUrl: String = "{{{basePath}}}") {
 
 {{#operation}}
 {{#javadocRenderer}}
 {{>javadoc}}
 {{/javadocRenderer}}
-  def {{operationId}}({{>methodParameters}}): ApiRequestT[{{>operationReturnType}}] =
+  def {{operationId}}({{>methodParameters}}): Request[{{#separateErrorChannel}}Either[ResponseError[Exception], {{>operationReturnType}}]{{/separateErrorChannel}}{{^separateErrorChannel}}{{>operationReturnType}}{{/separateErrorChannel}}, Nothing] =
     basicRequest
       .method(Method.{{httpMethod.toUpperCase}}, uri"$baseUrl{{{path}}}{{#queryParams.0}}?{{#queryParams}}{{baseName}}=${{{paramName}}}{{^-last}}&{{/-last}}{{/queryParams}}{{/queryParams.0}}{{#isApiKey}}{{#isKeyInQuery}}{{^queryParams.0}}?{{/queryParams.0}}{{#queryParams.0}}&{{/queryParams.0}}{{keyParamName}}=${apiKey.value}&{{/isKeyInQuery}}{{/isApiKey}}")
       .contentType({{#consumes.0}}"{{{mediaType}}}"{{/consumes.0}}{{^consumes}}"application/json"{{/consumes}}){{#headerParams}}
       .header({{>paramCreation}}){{/headerParams}}{{#authMethods}}{{#isBasic}}{{#isBasicBasic}}
-      .auth.withCredentials(basicAuth.user, basicAuth.password){{/isBasicBasic}}{{#isBasicBearer}}
-      .auth.bearer(bearerToken.token){{/isBasicBearer}}{{/isBasic}}{{#isApiKey}}{{#isKeyInHeader}}
-      .header("{{keyParamName}}", apiKey.value){{/isKeyInHeader}}{{#isKeyInCookie}}
-      .cookie("{{keyParamName}}", apiKey.value){{/isKeyInCookie}}{{/isApiKey}}{{/authMethods}}{{#formParams.0}}
+      .auth.withCredentials(username, password){{/isBasicBasic}}{{#isBasicBearer}}
+      .auth.bearer(bearerToken){{/isBasicBearer}}{{/isBasic}}{{#isApiKey}}{{#isKeyInHeader}}
+      .header("{{keyParamName}}", apiKey){{/isKeyInHeader}}{{#isKeyInCookie}}
+      .cookie("{{keyParamName}}", apiKey){{/isKeyInCookie}}{{/isApiKey}}{{/authMethods}}{{#formParams.0}}{{^isMultipart}}
       .body(Map({{#formParams}}
-        {{>paramFormCreation}},{{/formParams}}
-      )){{/formParams.0}}{{#bodyParam}}
+        {{>paramFormCreation}}{{#hasMore}}, {{/hasMore}}{{/formParams}}
+      )){{/isMultipart}}{{#isMultipart}}
+      .multipartBody(Seq({{#formParams}}
+        {{>paramMultipartCreation}}{{#hasMore}}, {{/hasMore}}{{/formParams}}
+      ).flatten){{/isMultipart}}{{/formParams.0}}{{#bodyParam}}
       .body({{paramName}}){{/bodyParam}}
-      .response(asJson[{{>operationReturnType}}])
+      .response({{#separateErrorChannel}}asJson{{/separateErrorChannel}}{{^separateErrorChannel}}asJsonAlwaysUnsafe{{/separateErrorChannel}}[{{>operationReturnType}}])
```

Comment:
Looks good. Could you help to separate api errors as exceptions as described here https://github.com/softwaremill/sttp-model/issues/5 ? The main idea is to split application exceptions (ideally with models) from transport level exceptions by using expected status codes.
Thanks

Reply:
Transport exceptions are already modeled using `SttpClientException`. These exceptions are thrown when read or connection exceptions occur, and result in a failed effect (failed `Future` or `IO`).

Though you are right that there's a piece missing here, if the OpenAPI description contains an e.g. JSON representation of an error? Then yes, we will need a custom hierarchy (as opposed to the standard `ResponseError`), covering three possibilities:

* `DeserializationError[T](body: String, error: T)` - if deserializing the success or error json fails
* `UnknownHttpError(body: String)` - for unmapped status codes
* `HttpError[T](body: T)` - for mapped status codes

Then, the overall request type would be: `Request[Either[ResponseError[T, U], V]]`, where:

* `T` is the type of deserialization errors, e.g. `io.circe.Error`
* `U` is the type of the http errors - but as there might be a couple of these, probably this needs to be a sealed trait, which is a implemented by all possible errors for this endpoint? This could be tricky, as each endpoint might have different error cases (e.g. `NotFound`).
* `V` is the type of the deserialized http success value (e.g. `User`)

Of course we'd need better naming for these :)

Reply:
That's always returned in the `Response` object which wraps the body. So the type of the result of a call is e.g. `Response[Either[ResponseError, V]]`, where `V` is the type of success